src/iop/toneequal.c

```diff
@@ -0,0 +1,2937 @@
+}
+
+static void quantization_callback(GtkWidget *slider, gpointer user_data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  if(self->dt->gui->reset) return;
+  dt_iop_toneequalizer_params_t *p = (dt_iop_toneequalizer_params_t *)self->params;
+
+  p->quantization = dt_bauhaus_slider_get(slider);
+  invalidate_luminance_cache(self);
+  dt_dev_add_history_item(darktable.develop, self, TRUE);
+}
+
+static void contrast_boost_callback(GtkWidget *slider, gpointer user_data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  if(self->dt->gui->reset) return;
+  dt_iop_toneequalizer_params_t *p = (dt_iop_toneequalizer_params_t *)self->params;
+
+  p->contrast_boost = dt_bauhaus_slider_get(slider);
+  invalidate_luminance_cache(self);
+  dt_dev_add_history_item(darktable.develop, self, TRUE);
+}
+
+static void exposure_boost_callback(GtkWidget *slider, gpointer user_data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  if(self->dt->gui->reset) return;
+  dt_iop_toneequalizer_params_t *p = (dt_iop_toneequalizer_params_t *)self->params;
+
+  p->exposure_boost = dt_bauhaus_slider_get(slider);
+  invalidate_luminance_cache(self);
+  dt_dev_add_history_item(darktable.develop, self, TRUE);
+}
+
+static void auto_adjust_exposure_boost(GtkWidget *quad, gpointer user_data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  if(self->dt->gui->reset) return;
+
+  dt_iop_request_focus(self);
+
+  if(!self->enabled)
+  {
+    // If module disabled, enable and do nothing
+    dt_dev_add_history_item(darktable.develop, self, TRUE);
+    return;
+  }
+
+  dt_iop_toneequalizer_params_t *p = (dt_iop_toneequalizer_params_t *)self->params;
+  dt_iop_toneequalizer_gui_data_t *g = (dt_iop_toneequalizer_gui_data_t *)self->gui_data;
+
+  if(!g->luminance_valid || self->dev->pipe->processing)
+  {
+    dt_control_log(_("wait for the preview to finish recomputing"));
+    return;
+  }
+
+  // The goal is to get the exposure distribution centered on the equalizer view
+  // to spread it over as many nodes as possible for better exposure control.
+  // Controls nodes are between -8 and 0 EV,
+  // so we aim at centering the exposure distribution on -4 EV
+  const float target = log2f(CONTRAST_FULCRUM);
+  update_histogram(g);
+  update_histogram_stats(g);
+  p->exposure_boost += target - g->histogram_average;
+
+  // Update the GUI stuff
+  const int reset = self->dt->gui->reset;
+  self->dt->gui->reset = 1;
+  dt_bauhaus_slider_set_soft(g->exposure_boost, p->exposure_boost);
+  self->dt->gui->reset = reset;
+  invalidate_luminance_cache(self);
+  dt_dev_add_history_item(darktable.develop, self, TRUE);
+}
+
+
+static void auto_adjust_contrast_boost(GtkWidget *quad, gpointer user_data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  if(self->dt->gui->reset) return;
+
+  dt_iop_request_focus(self);
+
+  if(!self->enabled)
+  {
+    // If module disabled, enable and do nothing
+  
```

Comment:
g->quantization   ->   g-bar

Reply:
Oh. :-) Yep, that fixed the crashes as well as the sliders reaction to single clicks.

Reply:
Ah, I came to the same conclusion in parallel :-) thanks